
# How to use dnpatterntools in Galaxy
## Introduction
 Nucleosomes provide measures of packaging and stabilize negative supercoiling of DNA in vivo; provide epigenetic layer of information guiding interactions of trans-acting proteins with the genome through their histone modifications; by their positioning they directly regulate access to the functional elements of the genome. Significant information about the occurrence of a nucleosome along DNA is encoded in certain features of the sequence (Lowary and Widom, 1998). Nucleosome-favoring DNA sequences are characterized by specific 10-10.4 base pairs periodical compositions of AA/TT and CC/GG dinucleotides (Ioshikhes et al., 2011 and references therein). The sequence features  together with other factors such as transcription factor binding and remodeling complexes play role in nucleosome positioning in vivo. 

 Nucleosomal DNAs are generally obtained from the puriﬁed chromatin stabilized with formaldehyde and digested with micrococcal nuclease (MNase) which cleaves sequence speciﬁc linker sites (Quintales, et al., 2015). To determine the nucleosome occupancy and positioning information from the MNase-Seq data, it is aligned to the reference genome of the organism under study. Then, nucleosomes positions and occupancy are determined from peaks resulting from the alignment and nucleosome-wrapped (nucleosomal) DNA  sequences can be extracted from those genomic locations.  

 The __dnpatterntools__ provide utilities to compute and analyze patterns of dinucleotide frequency distributions in a stack of nucleosome-wrapped DNA *fasta* sequences .

## Description of workflow
Computation of patterns of dinucleotide frequency distributions from nucleosome sequences consists of :
1. computation of distribution of frequency of dinucleotide occurrences in a stack of aligned sequences; 
2. determination of a statistical dyad position in sequences; 
3. obtaining statistical patterns of dinucleotide frequency of occurrence in the sequence applying symmetrization  and computing composite dinucleotides WW/SS (W = A or T and S=C or G) and RR/YY (R=A or G and Y=C or T);
4. normalization and smoothing of the patterns and computing their periodograms.

## Data [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3813510.svg)](https://doi.org/10.5281/zenodo.3813510) 

The  sequences of mouse (mm9) nucleosomal DNA that will be used in this demo are available from Zenodo . We will use sequences in [controlm.fa.gz](https://doi.org/10.5281/zenodo.3813510) .

## Hands on demo
Start your dnpatterntools-galaxy instance locally or in [docker](https://hub.docker.com/r/erinija/dnpatterntools-galaxy). Log in and start a new history. Name the history as __Dnpatterntools demo__.  
### Upload fasta file to your history
 - Copy the data link `https://zenodo.org/record/3813510/files/controlm.fa.gz`
 - Open the Galaxy Upload Manager 
 - Choose `Paste/Fetch data` and paste the copied link into the place holder
 - Set __Type__ to `fasta.gz` and __Genome__ to `mm9`
 - Press `Start` and `Close`. 
 
 The data file will appear in your history. It will be a compressed file. To extract fasta: 
 - Go to `Edit attributes` and select  a `Convert` tab
 - Select `Convert compressed file to uncompressed` and press `Convert datatype` button. 
 
 The new uncompressed fasta file appear in your history. To rename the data:  
 - Go to `Edit attributes` and select an `Attributes` tab
 - Change __Name__ to *controlm.fa* and press `Save`. 
 The renamed data file will reappear in your history.   
### Compute dinucleotides frequencies of occurrence in fasta file 
In a stack of nucleosomal DNA sequences aligned by an experimental end a frequency of occurrence of each dinucleotide is computed at each position. Given a binary matrix of dinucleotde occurrences in sequences coded as 1 and else as 0, a frequency proﬁle is simply a sum of occurrences of the selected dinucleotide at every position along the sequence. The sum is normalized by the number of sequences. To compute dinucleotide frequencies:
- From the Tools panel in a `Dnpatterntools` section select `Dinucleotide frequencies` tool
- In the field __From Fasta__ select *controlm.fa* data
- In the field __Dinucleotides__ input all 16 dinucleotides separated by space (this is a default option)
- Press `Execute`
- Inspect the data 
- Rename the dataset to `Dinucleotide frequencies of control mouse`

If there is no `Visualize this data` (Histogram) icon in your data section, then you need to install visualization tools yourself. Login as administrator and install `charts` tool of owner `iuc`. 

All data generated by dnpatterntools can be best visualized using the __Line with focus (NVD3)__ visualization tool. With the __Line with focus (NVD3)__ visualization tool visualize computed frequency profiles of AA, TT, CC, GG dinucleotides. An expected graph is shown in Figure 1. In the Figure you see the sharp narrow peak on the left of the graph. This peak indicates a micrococcal nuclease cleavage site. It can serve as a proof that your fasta file indeed contains a nucleosomal DNA sequences aligned by an experimental end. Don't forget to save your veisalization in `Saved visualizations` of the `User` by pressing  `Save` button. 
![Fig1](https://github.com/erinijapranckeviciene/dnpatterntools/blob/master/tools-extra/tutimg/Fig1.PNG "Frequency profiles of dinucleotides")
__Figure 1. Frequency profiles of the AA, TT, CC, GG dinucleotides.__ 

### Determine a dyad and nucleosome's position in a stack of the nucleosomal DNA sequences aligned by an experimental end 
Symmetry is a hallmark of the nucleosome DNA (Luger et al., 1997) and peaks in dinucleotide frequency patterns along the nucleosomal sequence have a recognizable dyad-symmetry.  The AA/TT, AT, GC, CC/GG, RR/YY dinucleotides statistically are preferred and they are periodically distributed along nucleosome DNA sequence (Luger et al., 1997). Dyad-symmetry feature helps to determine statistically a position of a dyad and a nucleosome in a stack of sequences aligned by an experimental end. At the nucleosome position centered on dyad the frequency proﬁles of dinucleotides on forward and reverse complementary strands will have a maximum positive correlation. A Pearson correlation between the dinucleotide frequency proﬁles on forward and reverse complementary strand is computed at each position of the nucleosome sequence within a sliding window 146bp long. Because of varying nature of correlations across conditions , identification of the nucleosome position needs visual inspection to verify its proximity to the cleavage site. 

### in progress , will be completed shortly

### Select pattern interval, symmetrize and compute composite dinucleotide profiles
text
### Smooth and compute periodograms of the patterns 
text

## Conclusion
conlusion
